src/compute/src/logging/differential.rs

```diff
@@ -189,32 +156,121 @@ pub fn construct<A: Allocate>(
 
-    traces
+        traces
+    })
+}
+
+type Pusher<D> = Tee<Timestamp, (D, Timestamp, Diff)>;
+type OutputBuffer<'a, 'b, D> = ConsolidateBuffer<'a, 'b, Timestamp, D, Diff, Pusher<D>>;
+
+/// Bundled output buffers used by the demux operator.
+struct DemuxOutput<'a, 'b> {
+    batches: OutputBuffer<'a, 'b, usize>,
+    records: OutputBuffer<'a, 'b, usize>,
+    sharing: OutputBuffer<'a, 'b, usize>,
+}
+
+/// Event handler of the demux operator.
+struct DemuxHandler<'a, 'b, 'c> {
+    /// Demux output buffers.
+    output: &'a mut DemuxOutput<'b, 'c>,
+    /// The logging interval specifying the time granularity for the updates.
+    logging_interval_ms: u128,
+    /// The current event time.
+    time: Duration,
+    /// A capability usable for emitting outputs.
+    cap: &'a InputCapability<Timestamp>,
+}
+
+impl DemuxHandler<'_, '_, '_> {
+    /// Return the timestamp associated with the current event, based on the event time and the
+    /// logging interval.
+    fn ts(&self) -> Timestamp {
+        let time_ms = self.time.as_millis();
+        let interval = self.logging_interval_ms;
+        let rounded = (time_ms / interval + 1) * interval;
+        rounded.try_into().expect("must fit")
+    }
+
+    /// Handle the given differential event.
+    fn handle(&mut self, event: DifferentialEvent) {
+        use DifferentialEvent::*;
+
+        match event {
+            Batch(e) => self.handle_batch(e),
+            Merge(e) => self.handle_merge(e),
+            Drop(e) => self.handle_drop(e),
+            TraceShare(e) => self.handle_trace_share(e),
+            _ => (),
+        }
+    }
+
+    fn handle_batch(&mut self, event: BatchEvent) {
+        let ts = self.ts();
+        let op = event.operator;
+        self.output.batches.give(self.cap, (op, ts, 1));
+
+        let diff = Diff::try_from(event.length).expect("must fit");
+        self.output.records.give(self.cap, (op, ts, diff));
+    }
+
+    fn handle_merge(&mut self, event: MergeEvent) {
+        let Some(done) = event.complete else {
+            return;
+        };
+
+        let ts = self.ts();
+        let op = event.operator;
+        self.output.batches.give(self.cap, (op, ts, -1));
+
+        let diff = Diff::try_from(done).expect("must fit")
+            - Diff::try_from(event.length1 + event.length2).expect("must fit");
+        if diff != 0 {
+            self.output.records.give(self.cap, (op, ts, diff));
+        }
+    }
+
+    fn handle_drop(&mut self, event: DropEvent) {
+        let ts = self.ts();
+        let op = event.operator;
+        self.output.batches.give(self.cap, (op, ts, -1));
+
+        let diff = -Diff::try_from(event.length).expect("must fit");
+        if diff != 0 {
+            self.output.records.give(self.cap, (op, ts, diff));
+        }
+    }
+
+    fn handle_trace_share(&mut self, event: TraceShare) {
+        let ts = self.ts();
+        let op = event.operator;
+        let diff = Diff::cast_from(event.diff);
+        if diff != 0 {
```

Comment:
I changed that because I assumed we wouldn't want to crash here just because timely decided to emit useless events at some point in the future. And ignoring is also how we handle zero-diffs for the other event types. That said, I might be missing something so happy to await @antiguru's verdict!

Reply:
If it would be a bug, then shouldn't we leave some kind of assert, at least a `debug_assert`?